### PR TITLE
pubsub fix for subscription.

### PR DESCRIPTION
**Pull request by Amenocy**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -10,7 +10,7 @@ npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
 lerna-debug.log*
-
+package-lock.json
```
Comment (XavierGeerinck) [inline .gitignore]: Please remove again

File: examples/http/hello-world/package.json [hunk 2]
```diff
@@ -19,7 +19,7 @@
     "typescript": "^4.2.4"
   },
   "dependencies": {
-    "@roadwork/dapr-js-sdk": "file:../../../build",
```
Comment (XavierGeerinck) [inline examples/http/hello-world/package.json]: Please remove from this commit, let's make this a separate discussion

File: src/http/lib/pubsub.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import fetch from 'node-fetch';
 import WebServerSingleton from "./WebServer/WebServerSingleton";
-import { TypeDaprPubSub } from '../types/DaprPubSub.type';
+import { TypeDaprPubSub, TypeElementOfDaprPubSub } from '../types/DaprPubSub.type';
```
Comment (XavierGeerinck) [inline src/http/lib/pubsub.ts:3]: I would name this: TypeDaprPubSubSubscriptions

File: src/http/lib/pubsub.ts [hunk 4]
```diff
@@ -23,30 +23,34 @@ export default class DaprPubSub {
     return res.status;
   }
 
-  async subscribe(pubSubName: string, topic: string, cb: TypeDaprPubSub) {
+  async subscribe(subs: TypeElementOfDaprPubSub[]) {
     const server = await WebServerSingleton.getServer();
+    if (subs.length < 1) return;
```
Comment (XavierGeerinck) [inline src/http/lib/pubsub.ts]: Not required, forEach won't run on empty array, just make sure to add subs: TYPE = []

File: src/http/lib/pubsub.ts [hunk 5]
```diff
@@ -23,30 +23,34 @@ export default class DaprPubSub {
...
+    
+    subs.forEach((sub: TypeElementOfDaprPubSub, i) => {
```
Comment (XavierGeerinck) [inline src/http/lib/pubsub.ts]: Would love a Promise.all here, please do a .map and apply the .map to a Promise.all for parallel registration

File: src/http/lib/pubsub.ts [hunk 6]
```diff
@@ -23,30 +23,34 @@ export default class DaprPubSub {
...
-        {
-          pubsubname: pubSubName,
-          topic,
-          route: `route-${pubSubName}-${topic}`,
-        },
-      ]);
+      console.log(`[Dapr API][PubSub] Registering`);
+      let subsCopy : TypeElementOfDaprPubSub[] = JSON.parse(JSON.stringify(subs));
+      subsCopy = subsCopy.map((row:TypeElementOfDaprPubSub)=>{
+        row.cb = null;
```
Comment (XavierGeerinck) [inline src/http/lib/pubsub.ts:48]: I don't like this implementation tbh, can you think of a different way to do this?

File: tsconfig.json [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
     /* Basic Options */
     // "incremental": true,                         /* Enable incremental compilation */
-    "target": "es5",                                /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
+    "target": "es2016",                                /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
```
Comment (XavierGeerinck) [inline tsconfig.json]: Why is this changed?

File: src/http/types/DaprPubSub.type.ts [hunk 8]
```diff
@@ -1,3 +1,9 @@
 import { IRequest, IResponse } from "../lib/WebServer";
 
-export type TypeDaprPubSub = (req: IRequest, res: IResponse) => Promise<void>;
\ No newline at end of file
+export type TypeDaprPubSub = (req: IRequest, res: IResponse) => Promise<void>;
+export type TypeElementOfDaprPubSub = {
+    pubSubName:String,
```
Comment (XavierGeerinck) [inline src/http/types/DaprPubSub.type.ts]: code formatting please :) space between the : and String (e.g. `pubSubName: string`)

**Review by XavierGeerinck — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 7 above)
Comment (Amenocy) [reply tsconfig.json]: i had compile errors... forgot where

File: src/http/lib/pubsub.ts (hunk 6 above)
Comment (Amenocy) [reply src/http/lib/pubsub.ts:48]: it is just for removing cb from data, there is no problem at all sending cb to Dapr.

File: src/http/lib/pubsub.ts (hunk 3 above)
Comment (Amenocy) [reply src/http/lib/pubsub.ts:3]: but it is just for one element of the array. so I think that's okay. if you want I will change it later

File: src/http/lib/pubsub.ts (hunk 6 above)
Comment (XavierGeerinck) [reply src/http/lib/pubsub.ts:48]: Yes, but this should be simpler I think? e.g. https://stackoverflow.com/a/53443378

Comment (Amenocy) [reply src/http/lib/pubsub.ts:48]: object assign not work in this case bkz it is shallow copy and we need deep copy.

**Comment by XavierGeerinck** [comment]
Merged! Thanks for the contribution 😀
